In [1]:
# Question 1
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [11]:
# Question 2.1
print(df.shape)

(391, 11)


##### Question 2.2
Observations are specific units of data regarding what is being studied in a dataset. Variables are  categories of attributes that vary between different observations.

In [24]:
# Question 3
villagers_df = df

villagers_df.describe() # (copied from chatbot)

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [21]:
# Question 3 (cont.)
villagers_df['species'].value_counts() # (copied from chatbot)

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [21]:
# Question 4
titanic_df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")
print(titanic_df.shape)
titanic_df.describe()

(891, 15)


survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

##### Question 4 (cont.)
The size of the dataset given by `.shape` is 891 rows and 15 columns.

The discrepancy between the number of rows analyzed by `.describe()` and the value reported for the count under "age," 714 instead of 891, can be explained by many entries in the "age" column not containing any data and therefore not being counted by `.describe()`.

##### Question 5
Attributes are variables assigned to objects that hold data representing properties of the object.

Methods are functions that are members of a class that operate on instances of the class. They define the behaviour and actions of objects, and often perform operations with/on attributes.

##### Question 6

"count"
    The number of non-missing values in the dataset, and also the number of values used to calculate the other statistics.
    
"mean" (arithmetic mean)
    The average value.
    
"std" (standard deviation)
    The measure of the amount of variance of the values in relation to their mean.
    
"min" (minimum value)
    The smallest value.

"25%" (25th percentile)
    The value at which 25% of the values are smaller and 75% are greater.

"50%" (50th percentile)
    The value at which 50% of the values are smaller and 50% are greater.

"75%" (75th percentile)
    The value at which 75% of the values are smaller and 25% are greater.

"max" (maximum value)
    The greatest value.

##### Question 7.1
One use-case in which using `df.dropna()` instead of `del df['col']` might be preferred is in a dataset in which the missing elements are concentrated more densely within fewer rows than they are in the columns, as dropping all rows with missing data would result in less data being lost than if all columns with missing data were deleted. 

For example, in a dataset recording information about 1438 wild pangolins, such as age and weight, in which 4 specimens have almost no information recorded about them, it is better to drop the 4 rows instead of deleting several columns and the hundreds of observations contained in them.

##### Question 7.2
An opposite use-case would be when given a data-set in which missing data is concentrated more densely within the columns compared to the rows. For example, if the only missing data in the pangolin dataset was instead under the column recording their favourite food, with the favourite foods of only 7 pangolins being recorded, much less data would be lost if the single column was dropped instead of  the data for the other 1431 pangolins.

##### Question 7.3
It may be important to apply `del df['col']` before `df.dropna()` because `dropna()` drops all rows with missing data, including those whose only missing elements are contained in a column that is marked for deletion. Applying `dropna()` first may end up dropping rows in which the only empty elements are in the column that are about to be deleted from the dataset regardless.

In [65]:
# Question 7.4
titanic_df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

print(titanic_df.isna().sum())

initial_count = titanic_df.count().sum()
print("\nInitial count:\t\t" + str(initial_count))
print("# of missing items:\t" + str(titanic_df.isna().sum().sum()))

del titanic_df['age']
del titanic_df['deck']

titanic_df = titanic_df.dropna()

print("\nFinal count:\t\t" + str(titanic_df.count().sum()))
print("# of missing items:\t" + str(titanic_df.isna().sum().sum()))
print("\n# of items deleted:\t" + str(initial_count - titanic_df.count().sum()))

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Initial count:		12496
# of missing items:	869

Final count:		11557
# of missing items:	0

# of items deleted:	939


##### Question 7.4 (cont.)
I deleted the "age" and "deck" columns before deleting all rows with missing values due to the vast majority of the missing data being concentrated in those two columns, with only 4 other missing elements. This approach minimizes the amount of data lost compared to if I called `.dropna()` on the dataset first, which would delete hundreds of rows of near-complete data.

In [86]:
# Question 8.1
df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv")

# df.groupby("col1")["col2"].describe()
df.groupby("row_n")["id"].describe()


count unique       top freq
row_n                            
2         1      1   admiral    1
3         1      1   agent-s    1
4         1      1     agnes    1
6         1      1        al    1
7         1      1   alfonso    1
...     ...    ...       ...  ...
475       1      1    winnie    1
477       1      1  wolfgang    1
480       1      1      yuka    1
481       1      1      zell    1
483       1      1    zucker    1

[391 rows x 4 columns]

##### Question 8.1 (cont.)
`df.groupby("row_n")` creates a group of rows for every unique value in the first column, "row_n", and puts rows with the same values in the same groups. As every element under "row_n" is already unique, this effectively groups each row by itself.
`df.groupby("row_n")["id"]` selects only the second column, "id", after grouping by "row_n".
`df.groupby("row_n")["id"].describe()` then provides summary statistics on the selected "id" column in each group. Due to this column consisting of strings, it only provides 4 statistics.

This results in a dataframe of summary statistics for the strings under "id" within every group of column "row_n." However, because every item under "row_n" is a unique value, each group consists of one row, making the line of code `df.groupby("col1")["col2"].describe()` very impractical when used on this particular dataset.

In [1]:
# Question 8.3.A
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

NameError: name 'pd' is not defined

In [5]:
# Question 8.3.B
import pandas as pd
df = pd.read_csv("titanics.csv")
df.isna().sum()

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [6]:
# Question 8.3.C
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
dF.isna().sum()

NameError: name 'df' is not defined

In [8]:
# Question 8.3.D
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url
df.isna().sum()

SyntaxError: '(' was never closed (3771808236.py, line 4)

In [11]:
# Question 8.3.E
df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv")
df.groupby("row_n")["id"].decribe()

AttributeError: 'SeriesGroupBy' object has no attribute 'decribe'

In [12]:
# Question 8.3.F
df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv")
df.groupby("Row_n")["id"].describe()

KeyError: 'Row_n'

In [13]:
# Question 8.3.E (cont.)
df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv")
df.groupby("row_n")["iD"].describe()

KeyError: 'Column not found: iD'

In [14]:
# Question 8.3.G
df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv")
df.groupby("row_n")[id].describe()

KeyError: 'Column not found: <built-in function id>'

##### Question 9
Somewhat